Updating hub deployment information

choldgraf opened the pull request:
Updates the hub deployment information as well as the template we use to accomplish this, with the following goals:

1. De-duplicate some information
2. Provide more explanation for what fields mean, and why they're needed
3. Create more pointers between the issue template and docs
4. Re-organize the "add a new hub" docs so they're in nested sections, since they fall under the same conceptual topic.
5. Renames "Community Champion" to "Community Representative" since it is a more descriptive term.

closes https://github.com/2i2c-org/pilot-hubs/issues/527

choldgraf commented:
I suspect this needs another round or two of edits before merging, but putting this PR up there in case others have thoughts on how it could be better. Maybe in particular @sgibson91 since she recently went through the hub deploy process.

I'd love comments on:

- Whether this is the right place for this kind of information
- Whether it will be intuitive / useful to have this combination of "issue template + docs"
- Whether we're missing any important information

sgibson91 reviewed (commented):
Great start! I left a few comments. I know it will involve some duplication, but I think I'd like as much of the info in the docs _also_ available in the issue template/form. Simply because it means the Hub Representative won't have to navigate away to get context/understanding.

I guess I'm also blurring the lines between who is filling in all these details, an engineer or a Hub Representative? If it's an engineer (maybe following a discussion within the opened issue), I guess my above comment could be addressed by a link to the docs you've provided and a prompt to "go here to understand how to help our engineer fill this in" (or something catchier).

choldgraf commented on .github/ISSUE_TEMPLATE/new-hub.md:
```diff
@@ -1,37 +1,50 @@
+- **Hub ID**:
+- **Hub Cluster**:
+- **Hub url**:
+- **Hub template**:
```
Yeah that makes sense assuming that the Community Representative will be the one filling this in. I like framing the question around features ("Do you want XX capability?") rather than implementation ("Do you want to use template XX").

Though for now I think the only feature flag we'd likely add is "Do you want to use a scalable dedicated cluster for Dask?" (since that's the difference between "regular Dask" and "Dask Gateway", I think)

choldgraf commented on docs/howto/operate/new-hub/index.md:
```diff
@@ -0,0 +1,80 @@
+Hub ID
+: The unique ID used to identify this hub.
+  The Community Representative should provide a URL-friendly community identifier (e.g. `my-university`).
+Hub Cluster
+: The cloud cluster where this hub will be run.
+  For example, `pilot` or `cloudbank`.
```
Right so maybe we should add a section to the bottom for the 2i2c Engineer that is deploying the hub to fill in, that would have more 2i2c-specific info in there?

sgibson91 replied:
Yeah, I think so

choldgraf commented on docs/howto/operate/new-hub/index.md:
```diff
@@ -0,0 +1,80 @@
+: The cloud cluster where this hub will be run.
+  For example, `pilot` or `cloudbank`.
+Hub Template
+: The configuration template that defines the basic infrastructure setup for the hub.
+  There are two options:
+
+  1. `basehub`: a basic JupyterHub with a fairly standard setup.
+  2. `daskhub`: a JupyterHub with a Dask Gateway enabled for scalable computing.
```
What do you think about turning this into a GitHub Issue form? That way we could add extra explanations and such from within the form (https://docs.github.com/en/communities/using-templates-to-encourage-useful-issues-and-pull-requests/configuring-issue-templates-for-your-repository#creating-issue-forms).

If others think this is a reasonable thing to try, I can try to figure out the new forms syntax and put that together.

sgibson91 replied:
GitHub docs: https://docs.github.com/en/communities/using-templates-to-encourage-useful-issues-and-pull-requests/syntax-for-issue-forms

I also transformed some of my issue templates into the forms format too: https://github.com/sgibson91/.github/tree/main/.github/ISSUE_TEMPLATE

choldgraf commented on .github/ISSUE_TEMPLATE/new-hub.md:
```diff
@@ -1,37 +1,50 @@
+# Deployment Information
+_Information needed to make an initial deploy of the hub._
+
+- **Hub ID**:
+- **Hub Cluster**:
+- **Hub url**:
```
yeah good point - I think in this version of the template, there is information that is relevant to the community and to 2i2c engineers intermixed together. What do you think about my comment below regarding using:

- An issue form for the **Community Representative** to fill in
- The opened issue text has a few markdown fields that need to be filled in by a hub engineer (such as cluster name etc)

choldgraf reviewed (commented):
Oops I think only one of my comments is officially a part of this "review" - but lemme know what you think!

sgibson91 commented:
As a general response, I think splitting this into sections that are "to be filled in by the Community Representative" and "to be filled in by a 2i2c engineer" is a good thing. We need all that info in one place, but we need to separate what the Community will care about and what is implementation detail that can be at 2i2c's discretion.

choldgraf commented:
@sgibson91 whaddaya think about my fancy shmancy new GitHub form?!? :-)

https://github.com/2i2c-org/pilot-hubs/blob/new-hub/.github/ISSUE_TEMPLATE/new-hub.yml

sgibson91 commented on .github/ISSUE_TEMPLATE/new-hub.yml:
```diff
@@ -0,0 +1,164 @@
+      description: |
+        The main point-of-contact for this hub, and the person who requests changes to the hub on behalf of the community.
+        **This person will be the first Hub Administrator**, and may add new hub administrators from the JupyterHub UI.
+        Should be either a GitHub handle or email address, depending on the type of authentication you've indicated above.
```
```suggestion
        Should be either a GitHub handle or email address, depending on the type of authentication you've indicated below.
```
The auth field is actually below this one!

sgibson91 commented on .github/ISSUE_TEMPLATE/new-hub.yml:
```diff
@@ -0,0 +1,164 @@
+      value: |
+        # Ask for a new 2i2c JupyterHub!
+        
+        This template is designed to help you provide information about your 2i2c Hub so that a 2i2c Engineer can set it up properly.
+        Answer the sections below as best you can.
+        We can update information as we learn more.
```
While I generally approve of line break after every period, unfortunately the issue form doesn't render this as we've come to expect with Markdown. This bugs me on a personal level and shouldn't block this PR, but just raising it in case I'm not alone in my annoyance 😆

sgibson91 commented on .github/ISSUE_TEMPLATE/new-hub.yml:
```diff
@@ -0,0 +1,164 @@
+      description: |
+        A URL to an image that will be used on the hub's landing page. (optional)
+    validations:
+      required: true
```
```suggestion
      required: false
```
You've stated in the description that this is optional, but the issue form is expecting it to be required 😄

sgibson91 commented on .github/ISSUE_TEMPLATE/new-hub.yml:
```diff
@@ -0,0 +1,164 @@
+      description: |
+        The image registry service used for this hub's user image, if a custom user image is desired.
+        For example, a [quay.io registry image](https://quay.io/).
+        See [](customize/custom-image) for more information.
```
This link isn't rendering.

sgibson91 commented on .github/ISSUE_TEMPLATE/new-hub.yml:
```diff
@@ -0,0 +1,164 @@
+        ### Deployment information
+
+        **Hub ID:**
+          The unique ID used to identify this hub. Should be URL-friendly and short. (e.g. `my-university`).
```
```suggestion
          <!-- The unique ID used to identify this hub. Should be URL-friendly and short. (e.g. `my-university`). -->
```

I would make the descriptions for the fields in the Engineer section HTML comments, so it looks a bit neater when rendered.

sgibson91 reviewed (commented):
It's really cool! I've left some comments and little bug fixes :)

GeorgianaElena commented on .github/ISSUE_TEMPLATE/new-hub.yml:
```diff
@@ -0,0 +1,164 @@
+      label: Hub image
+      description: |
+        A name and tag for a user image, pointing to the registry in the field above.
+        A name and tag for a user image, pointing to the registry in the field above.  
```
This row is duplicated:
```suggestion
```

choldgraf commented:
OK I've updated the template per @sgibson91 and @GeorgianaElena's suggestions! I also added a section about preferred cloud provider to make sure that we don't miss that important piece of information if it is there. What do folks think? Is this a good-enough start to try it out with a potential community?

sgibson91 commented:
<img width="319" alt="Screenshot 2021-07-22 at 10 28 55" src="https://user-images.githubusercontent.com/44771837/126617898-362fc1eb-9126-49f7-acac-1b62fb341219.png">

I think there's some incorrect formatting with the dropdown selection, otherwise I'm +1!

choldgraf commented:
OK I think I fixed the dropdown (looks like we can't use markdown in those items, which makes sense since it's not really clickable). If you are +1 on it, then I'll merge and send this to a group which wants a hub for a hackathon in a week, and we can see how this goes.

choldgraf commented:
Will merge this one in so that we can iterate a bit :-)

damianavila commented:
I like the new form format 😉
